Send generic OTel spans to Weave Agents view (#2923)
wandb/docs: weave/guides/tracking/trace-agents-otel.mdx

## Authentication

You can send spans to the following endpoint:

* **Path**: `/agents/otel/v1/traces`
* **Method**: `POST`
* **Content-Type**: `application/x-protobuf`
* **Base URL**: `https://trace.wandb.ai` for Multi-tenant Cloud.

The endpoint accepts OTLP protobuf payloads only. Use a protobuf HTTP exporter such as `opentelemetry-exporter-otlp-proto-http` (Python) or `@opentelemetry/exporter-trace-otlp-proto` (TypeScript). The endpoint also accepts `gzip` and `deflate` content encoding.

To authenticate with the endpoint, set your [W&B API key](https://wandb.ai/settings) in your `OTLPSpanExporter` configuration using one of the following values:

- `wandb-api-key`: Your W&B API key as the value.
- `Authorization`: HTTP Basic authentication with `api` as the user and your W&B API key as the password. This form is useful when your exporter or collector only supports standard authorization headers.

## Project routing

Weave routes spans to a project using either OTel resource attributes or a request header:

- **Resource attributes** (recommended): Set `wandb.entity` to your W&B team or user name and `wandb.project` to the project name on your `TracerProvider` resource.

> - **`project_id` header**: Set the value to `<your-team-name>/<your-project-name>`.

Reviewer: the new style for replaceables is **Use bracket-style uppercase placeholders**, so i think we're updating to [YOUR-TEAM][YOUR-PROJECT]

|---|---|---|
| `invoke_agent` | One complete user-agent exchange. | A turn. |
| `chat` | One call to a language model API. | An LLM call within a turn. |
| `execute_tool` | One tool execution. | A tool call. |

Two more attributes control grouping:

- `gen_ai.conversation.id`: Groups turns into a conversation. Use a stable ID for the lifetime of the conversation.
- `gen_ai.agent.name`: Groups conversations under a named agent in the Agents tab.

Other GenAI semantic-convention attributes, such as `gen_ai.request.model`, `gen_ai.usage.input_tokens`, and `gen_ai.usage.output_tokens`, enrich the rendering with model names and token counts but are optional. For the agent data model that these spans map onto, see [Trace your agents](/weave/guides/tracking/trace-agents#the-agent-data-model).

## Configure with environment variables only

If your application or collector reads the standard OTel exporter environment variables, you can set the following variables and route spans to Weave without changing any code:

```bash lines
export OTEL_EXPORTER_OTLP_TRACES_ENDPOINT="https://trace.wandb.ai/agents/otel/v1/traces"
export OTEL_EXPORTER_OTLP_TRACES_HEADERS="wandb-api-key=$WANDB_API_KEY"

> export OTEL_RESOURCE_ATTRIBUTES="wandb.entity=<your-team-name>,wandb.project=<your-project-name>"

Reviewer: ```suggestion
export OTEL_RESOURCE_ATTRIBUTES="wandb.entity=[YOUR-TEAM],wandb.project=[YOUR-PROJECT]"
```

```

## Example: emit agent spans without the Weave SDK

The following example instruments a minimal agent turn using only OTel packages. It emits an `invoke_agent` span for the turn, a `chat` span for the LLM call, and an `execute_tool` span for a tool call, then exports them to the Weave agents endpoint.

First, install the required dependencies:

<CodeGroup>
```bash Python
pip install opentelemetry-sdk opentelemetry-exporter-otlp-proto-http
```

```bash TypeScript
npm install @opentelemetry/api @opentelemetry/sdk-trace-node @opentelemetry/sdk-trace-base @opentelemetry/resources @opentelemetry/exporter-trace-otlp-proto
```
</CodeGroup>

> Then set the `WANDB_API_KEY`, `ENTITY`, and `PROJECT` values in the following code, and then run it:

Reviewer: ```suggestion
Then set the `WANDB_API_KEY`, `[YOUR-TEAM]`, and `[YOUR-PROJECT]` values in the following code, and then run it:
```

<CodeGroup>
```python Python lines
import os
from opentelemetry import trace
from opentelemetry.exporter.otlp.proto.http.trace_exporter import OTLPSpanExporter
from opentelemetry.sdk import trace as trace_sdk
from opentelemetry.sdk.resources import Resource
from opentelemetry.sdk.trace.export import BatchSpanProcessor

> ENTITY = "<your-team-name>"

Reviewer: ```suggestion
ENTITY = "[YOUR-TEAM]"
```

> PROJECT = "<your-project-name>"

Reviewer: ```suggestion
PROJECT = "[YOUR-PROJECT]"
```

WEAVE_AGENTS_OTLP_ENDPOINT = "https://trace.wandb.ai/agents/otel/v1/traces"

# Create an API key at https://wandb.ai/settings

> WANDB_API_KEY = os.environ["WANDB_API_KEY"]

Reviewer: actually there are several WANDB_API_KEY around here and lines that follow, might make it slower to determine which one they replace, might be good to replace it here and up top as [YOUR-WANDB-API-KEY].  ? dunno
